src/Juvix/Compiler/Nockma/Translation/FromAsm.hs

```diff
@@ -0,0 +1,1048 @@
+module Juvix.Compiler.Nockma.Translation.FromAsm where
+
+import Juvix.Compiler.Asm.Data.InfoTable qualified as Asm
+import Juvix.Compiler.Nockma.Evaluator
+import Juvix.Compiler.Nockma.Pretty
+import Juvix.Compiler.Nockma.Stdlib
+import Juvix.Compiler.Pipeline.EntryPoint
+import Juvix.Prelude hiding (Atom, Path)
+
+type UserFunctionId = Symbol
+
+data FunctionId
+  = UserFunction UserFunctionId
+  | BuiltinFunction BuiltinFunctionId
+  deriving stock (Generic, Eq)
+
+instance Hashable FunctionId
+
+data BuiltinFunctionId
+  = BuiltinPow2Go
+  | BuiltinPow2
+  | BuiltinAppendRights
+  deriving stock (Eq, Enum, Bounded, Generic)
+
+instance Hashable BuiltinFunctionId
+
+newtype CompilerOptions = CompilerOptions
+  {_compilerOptionsEnableTrace :: Bool}
+
+fromEntryPoint :: EntryPoint -> CompilerOptions
+fromEntryPoint EntryPoint {..} =
+  CompilerOptions
+    { _compilerOptionsEnableTrace = _entryPointDebug
+    }
+
+data FunctionInfo = FunctionInfo
+  { _functionInfoPath :: Path,
+    _functionInfoArity :: Natural
+  }
+
+data CompilerCtx = CompilerCtx
+  { _compilerFunctionInfos :: HashMap FunctionId FunctionInfo,
+    _compilerConstructorArities :: ConstructorArities,
+    _compilerOptions :: CompilerOptions
+  }
+
+type ConstructorArities = HashMap Asm.Tag Natural
+
+type Offset = Natural
+
+data CompilerFunction = CompilerFunction
+  { _compilerFunctionName :: FunctionId,
+    _compilerFunctionArity :: Natural,
+    _compilerFunction :: Sem '[Compiler, Reader CompilerCtx] ()
+  }
+
+data StackId
+  = CurrentFunction
+  | ValueStack
+  | TempStack
+  | AuxStack
+  | FrameStack
+  | StandardLibrary
+  | FunctionsLibrary
+  deriving stock (Enum, Bounded, Eq, Show)
+
+-- | A closure has the following structure:
+-- [code totalArgsNum argsNum args], where
+-- 1. code is code to run when fully applied.
+-- 2. totalArgsNum is the number of arguments that the function
+--     which created the closure expects.
+-- 3. argsNum is the number of arguments that have been applied to the closure.
+-- 4. args is the list of args that have been applied.
+--    The length of the list should be argsNum.
```

Comment:
To enable fast append (with one replacement operation) we might want to store the path to the end of supplied arguments list

Reply:
Or we can store a list of argument sequences. So we have (in pseudo-Haskell data notation)
```
data Closure = Closure {
  code :: Nock,
  -- stored args
  args :: NonEmpty (PathToEnd, NonEmpty Value),
  -- number of expected arguments (to be still supplied to the closure - when it reaches 0 we do the call),
  argsNum :: Int
}
```
The extending the closure args is just list cons (we know the number of supplied arguments, so we can precompute PathToEnd at compile time):
```
args' = (PathToEndOfSuppliedArgs, suppliedArgs) : args
```
At the end transferring the closure args to function args requires going through the list recursively and appending:
```
flatten ((p, args) : as) = replace p args (flatten as)
```
I think we can't get around doing the appending at least once.

Closure extensions are rare, however, so we might want to have a data structure for the common special case:
```
data FlatClosure = FlatClosure {
  code :: Nock,
  -- path to end of stored args, for closure extension
  pathToEnd :: Path,
  -- stored args
  args :: NonEmpty Value,
  -- number of expected arguments (to be still supplied to the closure - when it reaches 0 we do the call),
  argsNum :: Int
}
```